### PR TITLE
hasManyFragments now defaults to an empty array

### LINKED ISSUE #40: Default value for `hasManyFragments`
Currently when a `defaultValue` option isn't specified for a `hasManyFragments` attribute, [it is set to `null`](https://github.com/lytics/ember-data.model-fragments/blob/dade27ee147902794dc837058b6f608496fd30e4/packages/model-fragments/lib/fragments/attributes.js#L241). While `null` is a valid value of the property, it is more intuitive that the default is an empty fragment array, and is more consistent with how `DS.hasMany` works already. ~~In fact there currently is no way to force the creation of a fragment array once a record has been created.~~ However this is a minor breaking change, so should probably slated for a future major version bump.

Refs #36.

EDIT: a fragment array property can be set explicitly to an array-like object, which can either be empty or contain valid fragment models.

### PR DESCRIPTION
**Pull request by christophersansone**
Fixes #40.

### CODE REVIEW

**Comment by christophersansone** [comment]
Love to get some feedback on the tests.  I couldn't quite tell which ones should be specifically testing for `null` and which ones should be testing for the default empty array.  I tend to think we should assume the default empty array in most cases and then test for the null default value as an edge case.

**Comment by slindberg** [comment]
Thanks for doing the work on this one :beers:

Since `null` is still a valid value for fragment arrays, those tests need to stick around. The rollback-to-empty-array case _should_ be covered by other tests, but it couldn't hurt to add an extra one.

File: packages/model-fragments/lib/fragments/attributes.js [hunk 1]
```diff
@@ -319,13 +319,15 @@ function getDefaultValue(record, options, type) {
 
   if (typeof options.defaultValue === "function") {
     value = options.defaultValue();
-  } else if (options.defaultValue) {
+  } else if ("defaultValue" in options) {
```
Comment (slindberg) [inline packages/model-fragments/lib/fragments/attributes.js]: I'm not a huge fan of `in`, can you change this to `options.defaultValue !== undefined`.

File: packages/model-fragments/tests/integration/fragments/dependent_state_test.js [hunk 2]
```diff
@@ -596,7 +596,7 @@ test("a `DS.hasManyFragments` fragment property that is set to null can be rolle
   });
 });
 
-test("a `DS.hasManyFragments` fragment property that is null can be rolled back", function() {
+test("a `DS.hasManyFragments` fragment property that is empty can be rolled back", function() {
```
Comment (slindberg) [inline packages/model-fragments/tests/integration/fragments/dependent_state_test.js]: These tests shouldn't change, since `null` is still valid.

**Comment by slindberg** [comment]
Also, since this will get merged as part of the 1.0.0 release, it should remove the warning about it changing and related tests in `deprecation_test.js`.

**Comment by christophersansone** [comment]
@slindberg Thanks for the feedback, I'll make those changes.  I originally made the changes before #126 was merged and became unsure how to handle it all afterwards.  I think what I'll do is remove the warning, remove `defaultValue: []` in the tests, arrange the tests to assume the empty array default, and then make sure there is test coverage specifically for `defaultValue: null` cases.  Agreed?

Also, I'm thinking it would help have a v1.0 branch.  As each PR for #110 is complete, it can be merged into v1.0 immediately.  If contributors just pull from master each time, we'll likely run into merge conflicts.

**Comment by christophersansone** [comment]
I made the changes as discussed above.

**Comment by christophersansone** [comment]
@slindberg I noticed a small inconsistency I thought I'd point out.  According to one of the tests, when a `hasManyFragments` property is initialized to `null`, then the value should remain `null` instead of using the default value.  This test was actually passing for the wrong reason: it was selecting the default value for any falsy value, but the default value was also null.

I made a change to `hasManyFragments` to use the default value only if the value is specifically `undefined`.  This will now allow nulls to remain null regardless of the default value.  However, this is not the case with `hasOneFragment`: any falsy value will use the default value.  Is that the expected behavior for both `hasOneFragment` and `hasManyFragments`?

Incidentally, I checked how Ember Data applies default values.  If the key exists in the data (using the `in` operator), it uses the value.  If the key does not exist, it uses the default value.  Personally, I'd prefer being as consistent with ED as possible: use the `in` operator and apply it to both `hasOneFragment` and `hasManyFragments`.  I'd be happy to make the change if you'd like.

**Comment by slindberg** [comment]
Sorry for the long delay! Summertime :smile: 

> This test was actually passing for the wrong reason: it was selecting the default value for any falsy value, but the default value was also null.

Good catch. To make sure we're talking about the same thing, you mean when a fragment property has a falsy value in the payload, the default value always clobbers it, correct?

> Is that the expected behavior for both hasOneFragment and hasManyFragments?

Nope, this is definitely a bug.

> Personally, I'd prefer being as consistent with ED as possible

Agreed. This should be fixed independent of this PR (then rebase this one on top of it), sounds like you're up for tackling that too?

**Comment by christophersansone** [comment]
@slindberg Ha, no worries for the delay, except that my availability has dried up!  I'm hoping to get to this by the end of the week.

**Comment by christophersansone** [comment]
Created #138 to address the default value issue.  PR coming shortly.
